pkg/controllers/allocation/batch.go

```diff
@@ -0,0 +1,155 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allocation
+
+import (
+	"sync"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+)
+
+const (
+	BatchCacheTTL             = 1 * time.Hour
+	BatchCacheCleanupInterval = 2 * time.Hour
+)
+
+type Batcher struct {
+	// MaxBatchPeriod is the maximum amount of time to batch incoming pods before flushing
+	MaxBatchPeriod time.Duration
+	// IdlePeriod is the amount of time to wait to flush a batch when there are no incoming pods but the batch is not empty
+	// It should be a smaller duration than MaxBatchPeriod
+	IdlePeriod time.Duration
+
+	// batches keeps a mapping of a key (like a provisioner name and namespace) to a specific Batch
+	batches *cache.Cache
+}
+
+// NewBatchManager creates a new batch manager to start multiple batch windows
+func NewBatcher(maxBatchPeriod time.Duration, idlePeriod time.Duration) *Batcher {
+	batchCache := cache.New(BatchCacheTTL, BatchCacheCleanupInterval)
+	batchCache.OnEvicted(func(key string, val interface{}) {
+		batch := val.(*Batch)
+		batch.close()
+	})
+	return &Batcher{
+		MaxBatchPeriod: maxBatchPeriod,
+		IdlePeriod:     idlePeriod,
+		batches:        batchCache,
+	}
+}
+
+// Add starts a batching window or updates an existing one based on a batching key
+func (m *Batcher) Add(key string) {
```

Comment:
Consider making these interfaces runtime.Object, and constructing the keys inside the implementation.

Reply:
Could you add docs here on whether this is concurrency-safe?